File: runme/mcgen/local/datan/datsrc/gincbt.c
#include <math.h>
#include "datsrc.h"

double LIBFUNC gincbt_(double const P_T *aa, double const P_T *bb,
               double const P_T *xx)
{
    /* System generated locals */
    double ret_val, d__1;

    /* Local variables */
    integer m;
    double d2m, d2m1, xlim, apl2m, a, b, x, cfnew, fnorm, a1, b1, a2, b2,
           cf, rm;
    logical reflec;

    xlim = (*aa + 1.) / (*aa + *bb + 1.);
    if (*xx < xlim) {
        reflec = FALSE_;
        a = *aa;
        b = *bb;
        x = *xx;
    } else {
        reflec = TRUE_;
        a = *bb;
        b = *aa;
        x = 1. - *xx;
    }
    if (x < 1e-8) {
/* function known at end of range */
        cf = 0.f;
    } else {
/* continued fraction */
        a1 = 1.;
        b1 = 1.;
        a2 = 1.;
        b2 = 1. - (a + b) * x / (a + 1.);
        fnorm = 1. / b2;
        cf = a2 * fnorm;
        for (m = 1; m <= 100; ++m) {
            rm = (double) m;
            apl2m = a + rm * 2.;
            d2m = rm * (b - rm) * x / ((apl2m - 1.) * apl2m);
            d2m1 = -(a + rm) * (a + b + rm) * x / (apl2m * (apl2m + 1));
            a1 = (a2 + d2m * a1) * fnorm;
            b1 = (b2 + d2m * b1) * fnorm;
            a2 = a1 + d2m1 * a2 * fnorm;
            b2 = b1 + d2m1 * b2 * fnorm;
            if (b2 != 0.f) {
/* renormalize and test for convergence */
                fnorm = 1. / b2;
                cfnew = a2 * fnorm;
                if ((d__1 = cf - cfnew, abs(d__1)) / cf < 1e-8) {
                    goto L20;
                }
                cf = cfnew;
            }
        }
L20:
        d__1 = 1. - x;
        cf = cf * pow(x, a) * pow(d__1, b) / (a * gbetaf_(&a, &b));
    }
    if (reflec) {
        ret_val = 1. - cf;
    } else {
        ret_val = cf;
    }
    return ret_val;
} /* gincbt_ */

File: runme/mcgen/local/datan/datsrc/gincbt.f
      DOUBLE PRECISION FUNCTION GINCBT(AA,BB,XX)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      LOGICAL REFLEC
      PARAMETER(EPSILN=1.D-8, ONE=1.D0, ZERO=0.D0, TWO=2.D0)
      XLIM=(AA+ONE)/(AA+BB+ONE)
      IF (XX.LT.XLIM) THEN
        REFLEC=.FALSE.
        A=AA
        B=BB
        X=XX
      ELSE
        REFLEC=.TRUE.
        A=BB
        B=AA
        X=ONE-XX
      END IF
      IF(X.LT.EPSILN) THEN
C function known at end of range
        CF=0.
      ELSE
C continued fraction
        A1=ONE
        B1=ONE
        A2=ONE
        B2=ONE-(A+B)*X/(A+ONE)
        FNORM=ONE/B2
        CF=A2*FNORM
        DO 10 M=1,100
          RM=DBLE(M)
          APL2M=A+TWO*RM
          D2M=RM*(B-RM)*X/((APL2M-ONE)*APL2M)
          D2M1=-(A+RM)*(A+B+RM)*X/(APL2M*(APL2M+1))
          A1=(A2+D2M*A1)*FNORM
          B1=(B2+D2M*B1)*FNORM
          A2=A1+D2M1*A2*FNORM
          B2=B1+D2M1*B2*FNORM
          IF(B2.NE.0.) THEN
C renormalize and test for convergence
            FNORM=ONE/B2
            CFNEW=A2*FNORM
            IF(ABS(CF-CFNEW)/CF .LT. EPSILN) GO TO 20
            CF=CFNEW
          END IF
   10   CONTINUE
   20   CF=CF*(X**A)*((ONE-X)**B)/(A*GBETAF(A,B))
      END IF
      IF(REFLEC) THEN
        GINCBT=ONE-CF
      ELSE
        GINCBT=CF
      END IF
      END
